content/blog/2020-12-09-slp/index.md

```diff
@@ -0,0 +1,184 @@
++++
+title = "Exploiting Superword Level Parallelism with Multimedia Instruction Sets"
+[extra]
+bio = """
+ Ankush Rayabhari is a first-year MS student in Computer Science at Cornell who
+ is interested in databases and compilers.
+"""
+[[extra.authors]]
+name = "Ankush Rayabhari"
+latex = true
++++
+
+## Background
+
+As applications process more and more data, processors now include so called 
+SIMD registers and instructions, to enable more parallelism. These registers are
+extra wide, a 512-bit wide register can hold 16 32-bit words. Instructions on 
+these registers perform the same operation on each of the packed data types.
+
+For example, on Intel processors, the instruction vaddps adds each of the 
+corresponding packed data elements together.
+```
+Instruction: vaddps zmm, zmm, zmm
+Operation:
+  FOR j := 0 to 15
+    i := j*32
+    dst[i+31:i] := a[i+31:i] + b[i+31:i]
+  ENDFOR
+```
+
+Just having these instructions doesn't mean applications can make use of them.
+Unless developers choose to handwrite assembly code or use libraries 
+that do so, it's up to the compiler to identify and exploit opportunities to use 
+these features. This paper proposes a new set of algorithms for that.
+
+## Previous/Related Approaches
+
+- **Vector parallelism:** At the time of this paper, existing vector compilers 
+  generally targeted loops  as they provide a natural opportunity to run the 
+  same operation on multiple pieces of data. They do this by using loop 
+  transformations. For example, by applying loop fission and scalar expansion, 
+  the following loop
+  ```
+  for (i=0; i<16; i++) {
+    localdiff = ref[i] - curr[i];
+    diff += abs(localdiff);
+  }
+  ```
+  can be transformed into
+  ```
+  for (i=0; i<16; i++) {
+    T[i] = ref[i] - curr[i];
+  }
+  for (i=0; i<16; i++) {
+    diff += abs(T[i]);
+  }
+  ```
+  While the first loop can use SIMD instructions to perform the subtraction of 
+  multiple iterations at once, the second loop can't.
+
+  Generalizations of this involve executing multiple iterations of a loop in 
+  parallel though use of multiple cores.
+
+- **Instruction level parallelism:** Here, multiple instructions are executed in 
+  a single cycle. SIMD instructions are one way of accomplishing this. Some 
+  processors also feature
+  [superscalar execution](https://en.wikipedia.org/wiki/Superscalar_processor) 
+  where rather than have a single instruction execute, multiple instructions 
+  are executed each cycle. This is very similar to using SIMD instructions but 
```

Comment:
FWIW, "some processors" perhaps undersells it—superscalar processors are everywhere these days except for, like, embedded devices.

Reply:
Done.